packages/gatsby/lib/query-runner/query-watcher.js

```diff
@@ -0,0 +1,83 @@
+/***
+ * Jobs of this module
+ * - Maintain the list of components in the Redux store. So monitor new pages
+ *   and add/remove components.
+ * - Watch components for query changes and extract these and update the store.
+ * - Ensure all page queries are run as part of bootstrap and report back when
+ *   this is done
+ * - Whenever a query changes, re-run all pages that rely on this query.
+ ***/
+
+const _ = require("lodash")
+const chokidar = require("chokidar")
+
+const { store } = require("../redux/")
+const { boundActionCreators } = require("../redux/actions")
+const queryExtractor = require("./query-extractor")
+const queryRunner = require("./query-runner")
+
+const pageComponents = {}
+const doneInitialQueryRun = []
+
+const watcher = chokidar.watch()
+watcher.on("change", path => {
+  queryExtractor(path).then(query => {
+    // Check if the query has changed
+    if (query !== store.getState().pageComponents[path].query) {
+      boundActionCreators.setPageComponentQuery({
+        query,
+        componentPath: path,
+      })
+      runQueriesForComponent(path)
+    }
+  })
+})
+
+const debounceNewPages = _.debounce(() => {
+  store.dispatch({
+    type: `BOOTSTRAP_STAGE`,
```

Comment:
I don't really like the pattern `constants.js` (you have 3 files to update for 1 action… so yeah, you moving slower) my best compromise to have flexibility and maintabiliy is to merge actionsCreator & constants (an example into my [redux sandbox](https://github.com/fabien0102/connect4react/blob/master/src/actions/actions.ts))